### PR TITLE
Improve surface generation

### PR DESCRIPTION
**Pull request by brisvag**
This PR contains fixes/improvements on the `Surface` generation:

1. A simple heuristic is used to guess whether the manually annotated points are "flipped" (solves #17, part of #15). It works by calculating distances between the extrema of two adjacent rows of points, and assuming that the total distance will *increase* if the row is flipped the wrong way around. It works on the example from #15.
2. Addresses #18 by "downgrading" the given `order` when necessary. For example, if the user provides points with shape `(2, 3, 3)` (2 rows of 3 points each in 3 dimensions), and requests an order of `3`, the order will be downgraded to 1 for interpolating between rows, and to 2 for in-row interpolation. Warnings are emitted when this happens.
3. Fixes some small inconsistencies in `orientations` cause by non-orthogonal basis vectors. If the surface grid is sheared (i.e: basiacally always unless it's a perfect square) the grid splines are not perfectly perpendicular. This resulted in `x` and `y` vector that were a bit off, which in turn resulted in "wrong" final orientations. I added a step that fixes the `x` vector to be orthogonal to `y` and the calculated `z`. This in practice means that `y` vectors (the ones going across annotated rows, not along) are "exact", while the `x` vectors can be a bit off. The previous behaviour was undetermined, but probably somewhere in between (see https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.from_matrix.html#id2)
4. Be more accurate when generating consistent spacing in the grid. Before, if the `separation` was relatively big compared to the full length of the annotation, you would end up with wrongly spaced points because of approximations. The logic was basically "does it fit 5 points? No? Then use 4", but they would be unequally spaced in each different-legnth row (to ensure a perfect NxM grid). Now the same logic happens, but oversampled by 10 (and then we only use 1 point every 10). In practice, this still ensures a NxM grid, while also dividing by 10 the spacing imprecision. The downside is that the grid will no longer cover the *full* annotation if it can't fit exactly.

I also removed some old unused code from the utils (the non-nan padding).

### CODE REVIEW

File: src/morphosamplers/surface_spline.py [hunk 1]
```diff
@@ -87,17 +113,24 @@ def _fix_spline_edges(self, splines, separation, order):
...
                 spline.sample(u, derivative_order=1)
                 for spline, u in zip(splines, us)
             ]
 
+            # flip directions of annotations if necessary
+            directions = estimate_point_row_directions(points)
+            # coordinates can be simply inverted
+            points = [pts[::dir] for pts, dir in zip(points, directions)]
+            # directions need to be inverted *and* flipped (since they are a vector)
+            derivatives = [der[::dir] * dir for der, dir in zip(derivatives, directions)]
```
Comment (alisterburt) [inline src/morphosamplers/surface_spline.py:126]: der der dir der dir der 😆

**Review by alisterburt — APPROVED** [review]
Looks good, thanks for this! Merging :)

**Comment by alisterburt** [comment]
there's actually a small test failure because invalid points now warns rather than errors - would you mind reworking the test before this is merged?

**Comment by alisterburt** [comment]
closes #15 #18

**Comment by brisvag** [comment]
Wait for the merge, I found out just earlier that my heuristic is flawed in some cases. I need to think a bit on this.

**Comment by alisterburt** [comment]
Calculating distance of just the first and last pairs of points is probably more robust? 🙂

**Comment by brisvag** [comment]
You might be right xD

**Comment by alisterburt** [comment]
cool! glad you could tidy this up nicely - merging!

**Pull request merged**